// @ionic/core/dist/collection/components/app/app.js
import { getIonMode } from '../../global/ionic-global';
import { rIC } from '../../utils/helpers';
import { isPlatform } from '../../utils/platform';
export class App {
    componentDidLoad() {
        rIC(() => {
            const { win, config } = this;
            if (!config.getBoolean('_testing')) {
                importTapClick(win, config);
            }
            importInputShims(win, config);
            importStatusTap(win, config);
            importHardwareBackButton(win, config);
            importFocusVisible(win);
        });
    }
    hostData() {
        const mode = getIonMode(this);
        return {
            class: {
                [`${mode}`]: true,
                'ion-page': true,
                'force-statusbar-padding': this.config.getBoolean('_forceStatusbarPadding')
            }
        };
    }
    static get is() { return "ion-app"; }
    static get originalStyleUrls() { return {
        "$": ["app.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["app.css"]
    }; }
    static get contextProps() { return [{
            "name": "win",
            "context": "window"
        }, {
            "name": "config",
            "context": "config"
        }]; }
    static get elementRef() { return "el"; }
}
function importHardwareBackButton(win, config) {
    const hardwareBackConfig = config.getBoolean('hardwareBackButton', isPlatform(win, 'hybrid'));
    if (hardwareBackConfig) {
        import('../../utils/hardware-back-button').then(module => module.startHardwareBackButton(win));
    }
}
function importStatusTap(win, config) {
    const statusTap = config.getBoolean('statusTap', isPlatform(win, 'hybrid'));
    if (statusTap) {
        import('../../utils/status-tap').then(module => module.startStatusTap(win));
    }
}
function importFocusVisible(win) {
    import('../../utils/focus-visible').then(module => module.startFocusVisible(win.document));
}
function importTapClick(win, config) {
    import('../../utils/tap-click').then(module => module.startTapClick(win.document, config));
}
function importInputShims(win, config) {
    const inputShims = config.getBoolean('inputShims', needInputShims(win));
    if (inputShims) {
        import('../../utils/input-shims/input-shims').then(module => module.startInputShims(win.document, config));
    }
}
function needInputShims(win) {
    return isPlatform(win, 'ios') && isPlatform(win, 'mobile');
}
