// @ionic/core/dist/cjs/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const __chunk_1 = require('./ionic-e1db78f3.js');
const __chunk_2 = require('./chunk-e9c8ab68.js');
const __chunk_3 = require('./chunk-6d230a70.js');

function setupConfig(config) {
    const win = window;
    const Ionic = win.Ionic;
    if (Ionic && Ionic.config && Ionic.config.constructor.name !== 'Object') {
        console.error('ionic config was already initialized');
        return;
    }
    win.Ionic = win.Ionic || {};
    win.Ionic.config = Object.assign({}, win.Ionic.config, config);
    return win.Ionic.config;
}

exports.PLATFORMS_MAP = __chunk_1.PLATFORMS_MAP;
exports.getPlatforms = __chunk_1.getPlatforms;
exports.isPlatform = __chunk_1.isPlatform;
exports.setupPlatforms = __chunk_1.setupPlatforms;
exports.testUserAgent = __chunk_1.testUserAgent;
exports.LIFECYCLE_DID_ENTER = __chunk_2.LIFECYCLE_DID_ENTER;
exports.LIFECYCLE_DID_LEAVE = __chunk_2.LIFECYCLE_DID_LEAVE;
exports.LIFECYCLE_WILL_ENTER = __chunk_2.LIFECYCLE_WILL_ENTER;
exports.LIFECYCLE_WILL_LEAVE = __chunk_2.LIFECYCLE_WILL_LEAVE;
exports.LIFECYCLE_WILL_UNLOAD = __chunk_2.LIFECYCLE_WILL_UNLOAD;
exports.actionSheetController = __chunk_3.actionSheetController;
exports.alertController = __chunk_3.alertController;
exports.loadingController = __chunk_3.loadingController;
exports.modalController = __chunk_3.modalController;
exports.pickerController = __chunk_3.pickerController;
exports.popoverController = __chunk_3.popoverController;
exports.toastController = __chunk_3.toastController;
exports.setupConfig = setupConfig;
