// @ionic/core/dist/cjs/ion-popover-controller.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const __chunk_1 = require('./ionic-e1db78f3.js');
const __chunk_3 = require('./chunk-6d230a70.js');

class PopoverController {
    constructor(hostRef) {
        __chunk_1.registerInstance(this, hostRef);
        this.doc = __chunk_1.getContext(this, "document");
    }
    /**
     * Create a popover overlay with popover options.
     *
     * @param options The options to use to create the popover.
     */
    create(options) {
        return __chunk_3.createOverlay('ion-popover', options);
    }
    /**
     * Dismiss the open popover overlay.
     *
     * @param data Any data to emit in the dismiss events.
     * @param role The role of the element that is dismissing the popover.
     * This can be useful in a button handler for determining which button was
     * clicked to dismiss the popover.
     * Some examples include: ``"cancel"`, `"destructive"`, "selected"`, and `"backdrop"`.
     * @param id The id of the popover to dismiss. If an id is not provided, it will dismiss the most recently opened popover.
     */
    dismiss(data, role, id) {
        return __chunk_3.dismissOverlay(document, data, role, 'ion-popover', id);
    }
    /**
     * Get the most recently opened popover overlay.
     */
    async getTop() {
        return __chunk_3.getOverlay(document, 'ion-popover');
    }
}

exports.ion_popover_controller = PopoverController;
