// @ionic/core/dist/cjs/ion-modal-controller.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const __chunk_1 = require('./ionic-e1db78f3.js');
const __chunk_3 = require('./chunk-6d230a70.js');

class ModalController {
    constructor(hostRef) {
        __chunk_1.registerInstance(this, hostRef);
        this.doc = __chunk_1.getContext(this, "document");
    }
    /**
     * Create a modal overlay with modal options.
     *
     * @param options The options to use to create the modal.
     */
    create(options) {
        return __chunk_3.createOverlay('ion-modal', options);
    }
    /**
     * Dismiss the open modal overlay.
     *
     * @param data Any data to emit in the dismiss events.
     * @param role The role of the element that is dismissing the modal.
     * This can be useful in a button handler for determining which button was
     * clicked to dismiss the modal.
     * Some examples include: ``"cancel"`, `"destructive"`, "selected"`, and `"backdrop"`.
     * @param id The id of the modal to dismiss. If an id is not provided, it will dismiss the most recently opened modal.
     */
    dismiss(data, role, id) {
        return __chunk_3.dismissOverlay(document, data, role, 'ion-modal', id);
    }
    /**
     * Get the most recently opened modal overlay.
     */
    async getTop() {
        return __chunk_3.getOverlay(document, 'ion-modal');
    }
}

exports.ion_modal_controller = ModalController;
