// @ionic/core/dist/cjs/ion-alert-controller.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const __chunk_1 = require('./ionic-e1db78f3.js');
const __chunk_3 = require('./chunk-6d230a70.js');

class AlertController {
    constructor(hostRef) {
        __chunk_1.registerInstance(this, hostRef);
        this.doc = __chunk_1.getContext(this, "document");
    }
    /**
     * Create an alert overlay with alert options.
     *
     * @param options The options to use to create the alert.
     */
    create(options) {
        return __chunk_3.createOverlay('ion-alert', options);
    }
    /**
     * Dismiss the open alert overlay.
     *
     * @param data Any data to emit in the dismiss events.
     * @param role The role of the element that is dismissing the alert.
     * This can be useful in a button handler for determining which button was
     * clicked to dismiss the alert.
     * Some examples include: ``"cancel"`, `"destructive"`, "selected"`, and `"backdrop"`.
     * @param id The id of the alert to dismiss. If an id is not provided, it will dismiss the most recently opened alert.
     */
    dismiss(data, role, id) {
        return __chunk_3.dismissOverlay(document, data, role, 'ion-alert', id);
    }
    /**
     * Get the most recently opened alert overlay.
     */
    async getTop() {
        return __chunk_3.getOverlay(document, 'ion-alert');
    }
}

exports.ion_alert_controller = AlertController;
